Flaky test: TestLargeMessage timed out / hung

<details><summary>expand for details</summary>

```
?   	google.golang.org/grpc/test/grpc_testing	[no test files]
panic: test timed out after 7m0s
goroutine 4010 [running]:
testing.startAlarm.func1()
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/testing/testing.go:1023 +0x139
created by time.goFunc
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/time/sleep.go:170 +0x52
goroutine 1 [chan receive, 4 minutes]:
testing.(*T).Run(0xc4203fe750, 0x9d0c62, 0x10, 0x9e3af8, 0xc420057c01)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/testing/testing.go:698 +0x582
testing.runTests.func1(0xc4203fe750)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/testing/testing.go:882 +0xab
testing.tRunner(0xc4203fe750, 0xc420057d78)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/testing/testing.go:657 +0x108
testing.runTests(0xc4201134c0, 0xc2b360, 0x3c, 0x3c, 0xc30080)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/testing/testing.go:888 +0x4e1
testing.(*M).Run(0xc420057f28, 0xc4200e9ed0)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/testing/testing.go:822 +0x1c4
main.main()
	google.golang.org/grpc/transport/_test/_testmain.go:160 +0x210
goroutine 17 [syscall, 6 minutes, locked to thread]:
runtime.goexit()
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/runtime/asm_amd64.s:2197 +0x1
goroutine 3985 [IO wait, 4 minutes]:
net.runtime_pollWait(0x14c1fe5f2400, 0x72, 0xbee628)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/runtime/netpoll.go:164 +0x5e
net.(*pollDesc).wait(0xc4200963e8, 0x72, 0xbee628, 0x0)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/net/fd_poll_runtime.go:75 +0x5b
net.(*pollDesc).waitRead(0xc4200963e8, 0xffffffffffffffff, 0x0)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/net/fd_poll_runtime.go:80 +0x42
net.(*netFD).accept(0xc420096380, 0x0, 0xbf0b40, 0xc420278080)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/net/fd_unix.go:430 +0x2c0
net.(*TCPListener).accept(0xc42024c180, 0xc420346e30, 0x45cb6c, 0xc420346e08)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/net/tcpsock_posix.go:136 +0x51
net.(*TCPListener).Accept(0xc42024c180, 0xbf7bf0, 0xc420354000, 0xc420278060, 0x9e37e0)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/net/tcpsock.go:228 +0x50
google.golang.org/grpc/transport.(*server).start(0xc4202a4080, 0xc4203fe000, 0x0, 0xc4200a2090, 0x0)
	/home/travis/gopath/src/google.golang.org/grpc/transport/transport_test.go:272 +0x51a
created by google.golang.org/grpc/transport.setUpServerOnly
	/home/travis/gopath/src/google.golang.org/grpc/transport/transport_test.go:370 +0x11b
goroutine 4018 [select, 4 minutes]:
google.golang.org/grpc/transport.(*recvBufferReader).read(0xc42049bf40, 0xc420506000, 0x7ffe, 0x7ffe, 0x1, 0xc4203a5c88, 0x8e5f79)
	/home/travis/gopath/src/google.golang.org/grpc/transport/transport.go:142 +0x3de
google.golang.org/grpc/transport.(*recvBufferReader).Read(0xc42049bf40, 0xc420506000, 0x7ffe, 0x7ffe, 0x1, 0x0, 0x0)
	/home/travis/gopath/src/google.golang.org/grpc/transport/transport.go:131 +0x85
google.golang.org/grpc/transport.(*transportReader).Read(0xc42028b470, 0xc420506000, 0x7ffe, 0x7ffe, 0x1, 0x0, 0x0)
	/home/travis/gopath/src/google.golang.org/grpc/transport/transport.go:394 +0x74
io.ReadAtLeast(0xbf0200, 0xc42028b470, 0xc4204ee000, 0x1fffe, 0x1fffe, 0x1fffe, 0xc4203a5db8, 0x8e5f09, 0xc4200e06c0)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/io/io.go:307 +0xb7
io.ReadFull(0xbf0200, 0xc42028b470, 0xc4204ee000, 0x1fffe, 0x1fffe, 0x46b67c, 0x20000, 0x93dc40)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/io/io.go:325 +0x73
google.golang.org/grpc/transport.(*Stream).Read(0xc420290f00, 0xc4204ee000, 0x1fffe, 0x1fffe, 0x1fffe, 0x1fffe, 0x1bffe)
	/home/travis/gopath/src/google.golang.org/grpc/transport/transport.go:378 +0x169
google.golang.org/grpc/transport.TestLargeMessage.func1(0xc42022b910, 0xbf7dc0, 0xc4200e06c0, 0xc4202a64e0, 0xc4203fe000)
	/home/travis/gopath/src/google.golang.org/grpc/transport/transport_test.go:946 +0x393
created by google.golang.org/grpc/transport.TestLargeMessage
	/home/travis/gopath/src/google.golang.org/grpc/transport/transport_test.go:952 +0x1ec
goroutine 4005 [select, 4 minutes]:
google.golang.org/grpc/transport.(*controlBuffer).get(0xc420294040, 0x1, 0x0, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/google.golang.org/grpc/transport/controlbuf.go:293 +0x29a
google.golang.org/grpc/transport.(*loopyWriter).run(0xc4202a6120, 0x0, 0x0)
	/home/travis/gopath/src/google.golang.org/grpc/transport/controlbuf.go:380 +0xa5
google.golang.org/grpc/transport.newHTTP2Server.func2(0xc420354000)
	/home/travis/gopath/src/google.golang.org/grpc/transport/http2_server.go:276 +0x1bb
created by google.golang.org/grpc/transport.newHTTP2Server
	/home/travis/gopath/src/google.golang.org/grpc/transport/http2_server.go:281 +0x17a2
goroutine 3984 [semacquire, 4 minutes]:
sync.runtime_Semacquire(0xc42022b91c)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/runtime/sema.go:47 +0x34
sync.(*WaitGroup).Wait(0xc42022b910)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/sync/waitgroup.go:131 +0xc2
google.golang.org/grpc/transport.TestLargeMessage(0xc4203fe000)
	/home/travis/gopath/src/google.golang.org/grpc/transport/transport_test.go:954 +0x20e
testing.tRunner(0xc4203fe000, 0x9e3af8)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/testing/testing.go:657 +0x108
created by testing.(*T).Run
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/testing/testing.go:697 +0x544
goroutine 4006 [select, 4 minutes]:
google.golang.org/grpc/transport.(*http2Server).keepalive(0xc420354000)
	/home/travis/gopath/src/google.golang.org/grpc/transport/http2_server.go:870 +0x714
created by google.golang.org/grpc/transport.newHTTP2Server
	/home/travis/gopath/src/google.golang.org/grpc/transport/http2_server.go:282 +0x17c7
goroutine 3969 [select, 4 minutes]:
google.golang.org/grpc/transport.(*controlBuffer).get(0xc420434000, 0x1, 0x0, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/google.golang.org/grpc/transport/controlbuf.go:293 +0x29a
google.golang.org/grpc/transport.(*loopyWriter).run(0xc4202a66c0, 0x0, 0x0)
	/home/travis/gopath/src/google.golang.org/grpc/transport/controlbuf.go:380 +0xa5
google.golang.org/grpc/transport.newHTTP2Client.func3(0xc4200e06c0)
	/home/travis/gopath/src/google.golang.org/grpc/transport/http2_client.go:298 +0x106
created by google.golang.org/grpc/transport.newHTTP2Client
	/home/travis/gopath/src/google.golang.org/grpc/transport/http2_client.go:308 +0x15ee
goroutine 4019 [select, 4 minutes]:
google.golang.org/grpc/transport.(*recvBufferReader).read(0xc42049be50, 0xc420452000, 0x7ffe, 0x7ffe, 0x1, 0xc420047c88, 0x8e5f79)
	/home/travis/gopath/src/google.golang.org/grpc/transport/transport.go:142 +0x3de
google.golang.org/grpc/transport.(*recvBufferReader).Read(0xc42049be50, 0xc420452000, 0x7ffe, 0x7ffe, 0x1, 0x0, 0x0)
	/home/travis/gopath/src/google.golang.org/grpc/transport/transport.go:131 +0x85
google.golang.org/grpc/transport.(*transportReader).Read(0xc42028b2c0, 0xc420452000, 0x7ffe, 0x7ffe, 0x1, 0x0, 0x0)
	/home/travis/gopath/src/google.golang.org/grpc/transport/transport.go:394 +0x74
io.ReadAtLeast(0xbf0200, 0xc42028b2c0, 0xc42043a000, 0x1fffe, 0x1fffe, 0x1fffe, 0xc420047db8, 0x8e5f09, 0xc4200e06c0)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/io/io.go:307 +0xb7
io.ReadFull(0xbf0200, 0xc42028b2c0, 0xc42043a000, 0x1fffe, 0x1fffe, 0x46b67c, 0x20000, 0x93dc40)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/io/io.go:325 +0x73
google.golang.org/grpc/transport.(*Stream).Read(0xc420290e00, 0xc42043a000, 0x1fffe, 0x1fffe, 0x1fffe, 0x1fffe, 0x1bffe)
	/home/travis/gopath/src/google.golang.org/grpc/transport/transport.go:378 +0x169
google.golang.org/grpc/transport.TestLargeMessage.func1(0xc42022b910, 0xbf7dc0, 0xc4200e06c0, 0xc4202a64e0, 0xc4203fe000)
	/home/travis/gopath/src/google.golang.org/grpc/transport/transport_test.go:946 +0x393
created by google.golang.org/grpc/transport.TestLargeMessage
	/home/travis/gopath/src/google.golang.org/grpc/transport/transport_test.go:952 +0x1ec
goroutine 3968 [IO wait, 4 minutes]:
net.runtime_pollWait(0x14c1fe5f2640, 0x72, 0x4980ba)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/runtime/netpoll.go:164 +0x5e
net.(*pollDesc).wait(0xc420096688, 0x72, 0xbee628, 0xffffffffffffffff)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/net/fd_poll_runtime.go:75 +0x5b
net.(*pollDesc).waitRead(0xc420096688, 0xc42041c000, 0x8000)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/net/fd_poll_runtime.go:80 +0x42
net.(*netFD).Read(0xc420096620, 0xc42041c000, 0x8000, 0x8000, 0x0, 0xbf1ec0, 0xbee628)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/net/fd_unix.go:250 +0x22b
net.(*conn).Read(0xc42000e1b8, 0xc42041c000, 0x8000, 0x8000, 0x9, 0x46bc50, 0x46bc50)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/net/net.go:181 +0x97
net.(*TCPConn).Read(0xc42000e1b8, 0xc42041c000, 0x8000, 0x8000, 0x8, 0x46bc50, 0xc4203a8bb0)
	<autogenerated>:72 +0x73
bufio.(*Reader).Read(0xc4202a63c0, 0xc420094038, 0x9, 0x9, 0x20, 0x0, 0x42d3d0)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/bufio/bufio.go:213 +0x7e7
io.ReadAtLeast(0xbefd80, 0xc4202a63c0, 0xc420094038, 0x9, 0x9, 0x9, 0xc4203a8cd8, 0x46bdb1, 0xc4202883ac)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/io/io.go:307 +0xb7
io.ReadFull(0xbefd80, 0xc4202a63c0, 0xc420094038, 0x9, 0x9, 0xc4203a8d28, 0xc4203a8d38, 0xc4202c2000)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/io/io.go:325 +0x73
golang.org/x/net/http2.readFrameHeader(0xc420094038, 0x9, 0x9, 0xbefd80, 0xc4202a63c0, 0x0, 0xc400000000, 0x48663c, 0x48c6c3)
	/home/travis/gopath/src/golang.org/x/net/http2/frame.go:237 +0xa3
golang.org/x/net/http2.(*Framer).ReadFrame(0xc420094000, 0xc4202883a0, 0xc420288300, 0x0, 0x0)
	/home/travis/gopath/src/golang.org/x/net/http2/frame.go:492 +0x118
google.golang.org/grpc/transport.(*http2Client).reader(0xc4200e06c0)
	/home/travis/gopath/src/google.golang.org/grpc/transport/http2_client.go:1130 +0x20d
created by google.golang.org/grpc/transport.newHTTP2Client
	/home/travis/gopath/src/google.golang.org/grpc/transport/http2_client.go:265 +0x11e9
goroutine 4007 [IO wait, 4 minutes]:
net.runtime_pollWait(0x14c1fe5f2580, 0x72, 0x4980ba)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/runtime/netpoll.go:164 +0x5e
net.(*pollDesc).wait(0xc4200966f8, 0x72, 0xbee628, 0xffffffffffffffff)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/net/fd_poll_runtime.go:75 +0x5b
net.(*pollDesc).waitRead(0xc4200966f8, 0xc42050e000, 0x8000)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/net/fd_poll_runtime.go:80 +0x42
net.(*netFD).Read(0xc420096690, 0xc42050e000, 0x8000, 0x8000, 0x0, 0xbf1ec0, 0xbee628)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/net/fd_unix.go:250 +0x22b
net.(*conn).Read(0xc42024c1a0, 0xc42050e000, 0x8000, 0x8000, 0x9, 0x46bc50, 0x46bc50)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/net/net.go:181 +0x97
net.(*TCPConn).Read(0xc42024c1a0, 0xc42050e000, 0x8000, 0x8000, 0x4, 0x46bc50, 0xc420344bf8)
	<autogenerated>:72 +0x73
bufio.(*Reader).Read(0xc42029b5c0, 0xc420094498, 0x9, 0x9, 0x27fe228, 0xc42027e120, 0x434ff1)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/bufio/bufio.go:213 +0x7e7
io.ReadAtLeast(0xbefd80, 0xc42029b5c0, 0xc420094498, 0x9, 0x9, 0x9, 0xc4200b2b60, 0xc420344d88, 0x4293bc)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/io/io.go:307 +0xb7
io.ReadFull(0xbefd80, 0xc42029b5c0, 0xc420094498, 0x9, 0x9, 0x0, 0x0, 0xc420344d68)
	/home/travis/.gimme/versions/go1.8.7.linux.amd64/src/io/io.go:325 +0x73
golang.org/x/net/http2.readFrameHeader(0xc420094498, 0x9, 0x9, 0xbefd80, 0xc42029b5c0, 0x0, 0x0, 0x93a2c0, 0xc42027e120)
	/home/travis/gopath/src/golang.org/x/net/http2/frame.go:237 +0xa3
golang.org/x/net/http2.(*Framer).ReadFrame(0xc420094460, 0xc4202b4110, 0xc4202b4110, 0x0, 0x0)
	/home/travis/gopath/src/golang.org/x/net/http2/frame.go:492 +0x118
google.golang.org/grpc/transport.(*http2Server).HandleStreams(0xc420354000, 0xc420278060, 0x9e37e0)
	/home/travis/gopath/src/google.golang.org/grpc/transport/http2_server.go:428 +0xd6
created by google.golang.org/grpc/transport.(*server).start
	/home/travis/gopath/src/google.golang.org/grpc/transport/transport_test.go:342 +0x839
FAIL	google.golang.org/grpc/transport	420.030s
make: *** [testrace] Error 1
```

</details>

--- a/transport/controlbuf.go
+++ b/transport/controlbuf.go
@@ -84,6 +84,13 @@ func (il *itemList) isEmpty() bool {
 // the control buffer of transport. They represent different aspects of
 // control tasks, e.g., flow control, settings, streaming resetting, etc.
 
+// registerStream is used to register an incoming stream with loopy writer.
+type registerStream struct {
+	streamID uint32
+	wq       *writeQuota
+}
+
+// headerFrame is also used to register stream on the client-side.
 type headerFrame struct {
 	streamID   uint32
 	hf         []hpack.HeaderField
@@ -457,30 +464,39 @@ func (l *loopyWriter) incomingSettingsHandler(s *incomingSettings) error {
 	return l.framer.fr.WriteSettingsAck()
 }
 
+func (l *loopyWriter) registerStreamHandler(h *registerStream) error {
+	str := &outStream{
+		id:    h.streamID,
+		state: empty,
+		itl:   &itemList{},
+		wq:    h.wq,
+	}
+	l.estdStreams[h.streamID] = str
+	return nil
+}
+
 func (l *loopyWriter) headerHandler(h *headerFrame) error {
 	if l.side == serverSide {
-		if h.endStream { // Case 1.A: Server wants to close stream.
-			// Make sure it's not a trailers only response.
-			if str, ok := l.estdStreams[h.streamID]; ok {
-				if str.state != empty { // either active or waiting on stream quota.
-					// add it str's list of items.
-					str.itl.enqueue(h)
-					return nil
-				}
-			}
-			if err := l.writeHeader(h.streamID, h.endStream, h.hf, h.onWrite); err != nil {
-				return err
-			}
-			return l.cleanupStreamHandler(h.cleanup)
+		str, ok := l.estdStreams[h.streamID]
+		if !ok {
+			warningf("transport: loopy doesn't recognize the stream: %d", h.streamID)
+			return nil
 		}
-		// Case 1.B: Server is responding back with headers.
-		str := &outStream{
-			state: empty,
-			itl:   &itemList{},
-			wq:    h.wq,
+		// Case 1.A: Server is responding back with headers.
+		if !h.endStream {
+			return l.writeHeader(h.streamID, h.endStream, h.hf, h.onWrite)
 		}
-		l.estdStreams[h.streamID] = str
-		return l.writeHeader(h.streamID, h.endStream, h.hf, h.onWrite)
+		// else:  Case 1.B: Server wants to close stream.
+
+		if str.state != empty { // either active or waiting on stream quota.
+			// add it str's list of items.
+			str.itl.enqueue(h)
+			return nil
+		}
+		if err := l.writeHeader(h.streamID, h.endStream, h.hf, h.onWrite); err != nil {
+			return err
+		}
+		return l.cleanupStreamHandler(h.cleanup)
 	}
 	// Case 2: Client wants to originate stream.
 	str := &outStream{
@@ -639,6 +655,8 @@ func (l *loopyWriter) handle(i interface{}) error {
 		return l.outgoingSettingsHandler(i)
 	case *headerFrame:
 		return l.headerHandler(i)
+	case *registerStream:
+		return l.registerStreamHandler(i)
 	case *cleanupStream:
 		return l.cleanupStreamHandler(i)
 	case *incomingGoAway:

--- a/transport/http2_server.go
+++ b/transport/http2_server.go
@@ -415,6 +415,11 @@ func (t *http2Server) operateHeaders(frame *http2.MetaHeadersFrame, handle func(
 			t.updateWindow(s, uint32(n))
 		},
 	}
+	// Register the stream with loopy.
+	t.controlBuf.put(&registerStream{
+		streamID: s.id,
+		wq:       s.wq,
+	})
 	handle(s)
 	return
 }
@@ -732,7 +737,6 @@ func (t *http2Server) writeHeaderLocked(s *Stream) {
 		onWrite: func() {
 			atomic.StoreUint32(&t.resetPingStrikes, 1)
 		},
-		wq: s.wq,
 	})
 	if t.stats != nil {
 		// Note: WireLength is not set in outHeader.

